notes/plutus-report/costing.tex

```diff
@@ -0,0 +1,339 @@
+That way we can terminate evaluation if we would exceed the budget after running the operation, without actually running the operation.
+
+\paragraph{Time for builtin operations}
+The approach we take for Time is very empirical:
+\begin{enumerate}
+\item Micro-benchmark the builtin with arguments of varying sizes.
+\item Look at the resulting graphs and try to pick an appropriate linear model for the running time based on the argument sizes (e.g. ``linear in both arguments'').
+\item Use linear regression to infer the parameters for the model, these are the cost model parameters for this builtin.
+\end{enumerate}
+
+\paragraph{Space for builtin operations}
+The approach we take for Space is more deductive.
+Generally it is hard to measure memory allocation, especially when it occurs in a foreign library.
+So we largely rely on reasoning about or inspecting the libraries in question.
+Many perform no allocation, or a predictable amount of allocation.
+
+\subsubsection{Space and Time for machine steps}
+The CEK machine itself has many steps that it goes through, which correspond to handling e.g. application of user-defined functions.
+These also need to have costs associated: naive profiling suggests that the operation machine itself (as opposed to builtin functions) is usually >50\% of evaluation time.
+
+\paragraph{Time for machine steps}
+The approach we initially took for Time is empirical, matching the approach for builtins:
+\begin{enumerate}
+\item Assume that the execution time is linear in all the kinds of machine steps.
+\item Benchmark execution of a large number of varied programs that use all the execution steps.
+\item Use linear regression to infer the parameters for the model, these are the cost parameters for the machine steps.
+\end{enumerate}
+
+However, when we actually did this, we found that the number of machine steps of different kinds are almost perfectly correlated, such that there is little point trying to infer coefficients for them individually.
+So instead we fit a single-parameter linear model based only on the total number of steps.
+
+\paragraph{Space for machine steps}
+The approach we take for Space is more deductive.
+Generally we assume that machine steps incur memory usage in certain specific ways, such as creating machine frames or creating mappings in variable environments, and we handle these specifically.
+
+\subsubsection{Summary of overall calculation}
+To sum up, the cost for a program execution is:
+\begin{itemize}
+\item The initial Space cost of the AST.
+\item The initial Time cost for starting the machine (the intercept of the linear model for the machine steps).
+\item The Time and Space costs for each machine step that is taken (as determined by the coefficients of the linear model for the machine steps).
+\item The Time and Space costs for each builtin call (as determined by the linear models for each builtin operation).
+\end{itemize}
+
```

Comment:
Seems reasonable.  Now that I think about it, we may want non-uniform space costs for the machine steps since, eg, some of them create new frames on the stack and others don't. In fact, we can probably track the size of the stack completely, even accounting for deallocation.  I'll try to remember to think about that. We should at least see how big the stack gets in practice (JIRA issue at SCP-1985).

Reply:
I think they probably do have non-uniform memory costs, but I expect that the fact that number of each type of step is so correlated will again make it so we might as well fit a one-parameter model.